[AutoFill Debugging] Add a simple versioning system for text extraction outputs

whsieh opened the pull request:
#### 1044a1057b9527782a2206cb73cbc279dd5805e6
<pre>
[AutoFill Debugging] Add a simple versioning system for text extraction outputs
<a href="https://bugs.webkit.org/show_bug.cgi?id=302493">https://bugs.webkit.org/show_bug.cgi?id=302493</a>
<a href="https://rdar.apple.com/164668165">rdar://164668165</a>

Reviewed by Abrar Rahman Protyasha and Lily Spiniolas.

Add support for a user default, `WebKit2TextExtractionOutputVersion`, which can be used to restore
some older text extraction behaviors (such as the replacing `textFormControls` with `input` /
`textarea`); we additionally include this version number as the last line in text extraction output,
for debugging convenience.

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::TextExtractionAggregator::TextExtractionAggregator):
(WebKit::TextExtractionAggregator::~TextExtractionAggregator):
(WebKit::TextExtractionAggregator::useTagNameForTextFormControls const):
(WebKit::TextExtractionAggregator::addLineForVersionNumberIfNeeded):
(WebKit::TextExtractionAggregator::version const):
(WebKit::addPartsForItem):
* Source/WebKit/Shared/TextExtractionToStringConversion.h:
(WebKit::TextExtractionOptions::TextExtractionOptions):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _debugTextWithConfiguration:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/303032@main">https://commits.webkit.org/303032@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cb9f781a200c19e933cdd0e0f20ead119c70257

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130967 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3292 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41926 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138409 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/baec3c2e-c017-49af-af52-e8ce1e73791d) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3274 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3133 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99768 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c954999c-0796-49c7-855d-f3a552db05f8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133913 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2347 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117248 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80474 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/eb0f390c-b9ea-45ea-9795-256d4f7247ba) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2265 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35380 "Found 4 new test failures: inspector/animation/lifecycle-css-transition.html inspector/page/overrideSetting-ICECandidateFilteringEnabled.html js/ShadowRealm-iframe-detach.html webaudio/audiobuffersource-not-gced-until-ended.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81654 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110855 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35883 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140901 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3034 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2716 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108286 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3080 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113579 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108243 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27539 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2300 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32004 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56071 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3102 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66497 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2923 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3123 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3032 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
```diff
@@ -395,7 +413,10 @@ static void addPartsForItem(const TextExtraction::Item& item, std::optional<Node
+            if (aggregator.useTagNameForTextFormControls())
+                parts.append(item.nodeName.convertToASCIILowercase());
+            else
+                parts.append("textFormControl"_s);
```
nit (and entirely opinion-based):
```suggestion
            const auto name = aggregator.useTagNameForTextFormControls() ? item.nodeName.convertToASCIILowercase() : "textFormControl"_s;
            parts.append(name);
```

whsieh replied:
Sounds good! Condensed this down to one `append` call.

aprotyas reviewed (approved):
Depending on how many version iterations or how much feature granularity you need to support, I would maybe (eventually) consider mapping the version to some `OptionSet<TextExtractionFeature>`-like data internally. Of course you'd still expose only the version number to clients, but it might make this kind of reasoning simpler?

```cpp
    using Version = unsigned;
    HashMap<Version, OptionSet<TextExtractionFeature>> featureMatrix {
        { 2, { ... | ... | TextExtractionFeature::TagNameForTextFormControls } },
    };
    bool useTagNameForTextFormControls() const
    {
        return featureMatrix.getOptional(effectiveVersion().transform([](auto features) { return features.contains(TextExtractionFeature::TagNameForTextFormControls); })).value_or(false);
    }
```

Of course this runs the risk of over-engineering and maybe goes against the assumption that new versions will only ever be a feature superset of prior versions 😅

whsieh committed:
[AutoFill Debugging] Add a simple versioning system for text extraction outputs
https://bugs.webkit.org/show_bug.cgi?id=302493
rdar://164668165

Reviewed by Abrar Rahman Protyasha and Lily Spiniolas.

Add support for a user default, `WebKit2TextExtractionOutputVersion`, which can be used to restore
some older text extraction behaviors (such as the replacing `textFormControls` with `input` /
`textarea`); we additionally include this version number as the last line in text extraction output,
for debugging convenience.

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::TextExtractionAggregator::TextExtractionAggregator):
(WebKit::TextExtractionAggregator::~TextExtractionAggregator):
(WebKit::TextExtractionAggregator::useTagNameForTextFormControls const):
(WebKit::TextExtractionAggregator::addLineForVersionNumberIfNeeded):
(WebKit::TextExtractionAggregator::version const):
(WebKit::addPartsForItem):
* Source/WebKit/Shared/TextExtractionToStringConversion.h:
(WebKit::TextExtractionOptions::TextExtractionOptions):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _debugTextWithConfiguration:completionHandler:]):

Canonical link: https://commits.webkit.org/303032@main
